runge_kutta.py

```diff
@@ -0,0 +1,228 @@
+    """
+
+    :param system_to_solve: callable function that returns list of equations to solve
+    :param k0: initial values for optimizing variables
+    :param hxyz: initial conditions for the system
+    :return: list of roots (k_coeffs)
+    """
+    hxyz = (h, *xyz)
+    root = fsolve(system_to_solve, k0, args=hxyz)
+    return root
+
+
+def make_one_step(xyz, h):
+    """"
+    Makes one step with the step size = h
+    """
+    k_coeffs = solve_ksystem(system_to_solve=ksystem, xyz=xyz, h=h)
+    xyz_one_step = np.array([xyz[0] + (k_coeffs[0] + k_coeffs[2]) / 2,
+                             xyz[1] + (k_coeffs[2] + k_coeffs[3]) / 2,
+                             xyz[2] + (k_coeffs[4] + k_coeffs[5]) / 2
+                             ])
+    return xyz_one_step
+
+
+def make_two_steps(xyz: tuple, h: float):
+    """"
+    Makes two steps with the step size = h/2
+    """
+    # k system coeffs (roots) for step = h/2
+    h = h / 2
+    k_coeffs = solve_ksystem(system_to_solve=ksystem, xyz=xyz, h=h)
+
+    # values for xyz with one step = h/2
+    xyz_one_half_step = np.array([xyz[0] + (k_coeffs[0] + k_coeffs[2]) / 2,
+                                  xyz[1] + (k_coeffs[2] + k_coeffs[3]) / 2,
+                                  xyz[2] + (k_coeffs[4] + k_coeffs[5]) / 2
+                                  ])
+
+    # values for xyz with two steps of step = h/2
+    xyz_two_half_step = np.array([xyz_one_half_step[0] + (k_coeffs[0] + k_coeffs[2]) / 2,
+                                  xyz_one_half_step[1] + (k_coeffs[2] + k_coeffs[3]) / 2,
+                                  xyz_one_half_step[2] + (k_coeffs[4] + k_coeffs[5]) / 2
+                                  ])
+    return xyz_two_half_step
+
+
+def calculate_iter_errors(xyz_one_step: np.array, xyz_two_half_step: np.array) -> np.array:
+    """"
+    Calculates iteration errors for each coordinate (x,y,z)
+    """
+    iter_errors = np.array([np.abs((xyz_one_step[0] - xyz_two_half_step[0]) / 15),
+                            np.abs((xyz_one_step[1] - xyz_two_half_step[1]) / 15),
+                            np.abs((xyz_one_step[2] - xyz_two_half_step[2]) / 15)
+                            ])
+    return iter_errors
+
+
+def euclidian_norm(arr: np.array):
+    """"
+    Calculates Euclidian norm of errors vector: √(err_x^2 + err_y^2 + err_z^2)
+    """
+    eu_norm = np.sqrt(np.sum(np.power(arr, 2)))
+    return eu_norm
+
+def make_iteration():
+    pass
+
+
+def run(t_start, t_end, xyz0, h0: float = 0.1, epsilon: int = 10e-4):
+
+    xyz = np.ndarray(shape=(1, 3))
+    xyz[0] = xyz0
+    tn = np.array([t_start])
+
+    # arrays for component errors
+    errors = np.array([0, 0, 0])
+    err_x = np.array([0])
+    err_y = np.array([0])
+    err_z = np.array([0])
+
+    # array for errors norm
+    error_norms = np.array([0])
+
+    # array for step lengths
+    steps = np.array([h0])
+    h = h0
+
+    curr_iter = 0
+    while tn[curr_iter] <= t_end:
+        if curr_iter > 10000:
```

Comment:
magic number, should be input arg with default value

Reply:
done